tensorrt_llm/_torch/auto_deploy/transform/library/sync_tied_weights.py

```diff
@@ -0,0 +1,312 @@
+            # Existing deduplication handles both-exported case
+            continue
+
+        # Cross-boundary case: one exported, one not
+        cross_boundary_pairs.append((dst_key, src_key))
+
+        # Determine which is canonical (exported)
+        if src_exported:
+            canonical_keys.add(src_key)
+        else:
+            canonical_keys.add(dst_key)
+
+    return cross_boundary_pairs, canonical_keys
+
+
+def _sync_tied_weights(
+    mod: nn.Module,
+    cross_boundary_pairs: List[Tuple[str, str]],
+    canonical_keys: Set[str],
+) -> int:
+    """Sync tied weights by making non-canonical weights point to canonical weights.
+
+    This function should be called AFTER weights are loaded. It makes the non-exported
+    weight (e.g., lm_head.weight) point to the same tensor as the exported weight
+    (e.g., embed_tokens.weight).
+
+    Args:
+        mod: The root model with loaded weights.
+        cross_boundary_pairs: List of (dst_key, src_key) pairs with cross-boundary tying.
+        canonical_keys: Set of parameter keys that are canonical (exported).
+
+    Returns:
+        Number of weights successfully synced.
+    """
+    synced_count = 0
+    for dst_key, src_key in cross_boundary_pairs:
+        # Determine canonical vs redirect keys
+        if src_key in canonical_keys:
+            canonical_key = src_key
+            redirect_key = dst_key
+        else:
+            canonical_key = dst_key
+            redirect_key = src_key
+
+        try:
+            # Get the loaded canonical parameter
+            canonical_param = mod.get_parameter(canonical_key)
+
+            # Parse redirect key into module path and param name
+            parts = redirect_key.rsplit(".", 1)
+            if len(parts) > 1:
+                redirect_mod = mod.get_submodule(parts[0])
+                redirect_name = parts[1]
+            else:
+                redirect_mod = mod
+                redirect_name = parts[0]
+
+            # Remove from _parameters so it's not a registered parameter
+            # (prevents double-counting in state_dict, optimizer, etc.)
+            if redirect_name in redirect_mod._parameters:
+                del redirect_mod._parameters[redirect_name]
+
+            # Sync: make redirect point to the canonical tensor
+            setattr(redirect_mod, redirect_name, canonical_param)
+            ad_logger.info(f"Synced tied weight: {redirect_key} -> {canonical_key} (canonical)")
+            synced_count += 1
+        except Exception as e:
+            ad_logger.warning(f"Failed to sync tied weight {redirect_key} -> {canonical_key}: {e}")
+
+    return synced_count
+
+
+class SyncTiedWeightsConfig(TransformConfig):
+    """Configuration for the sync tied weights transform."""
+
+    pass  # No configuration options needed for now
+
+
+@TransformRegistry.register("sync_tied_weights")
+class SyncTiedWeights(BaseTransform):
+    """Sync tied weights that cross the export boundary.
```

Comment:
we should also check if we can just do this with load hooks during/before export.